network/gossip/libp2p/topology/graphsampler.go

```diff
@@ -0,0 +1,17 @@
+package topology
+
+import (
+	"github.com/onflow/flow-go/model/flow"
+)
+
+// GraphSampler is an interface type responsible to sample a connected graph out of a
+// identity list. It is a must-have feature that independent executions of this interface on different
+// nodes (with different seeds) create a connected graph.
+type ConnectedGraphSampler interface {
+	// SampleConnectedGraph receives two lists: `all` and `shouldHave`. It then samples a connected fanout
+	// for the caller that includes all members of the `shouldHave` set.
+	// Independent invocations of this method over different nodes must create a connected graph.
+	// The term fanout is the set of nodes that this instance should get connected to in order to create a
+	// connected graph.
+	SampleConnectedGraph(all flow.IdentityList, shouldHave flow.IdentityList) (flow.IdentityList, error)
```

Comment:
This function can perhaps be `SampleGraph(ids flow.IdentityList, fanout uint) (flow.IdentityList, error)`
Then the caller can make the graph connected if it chooses to do so by passing in a fanout of `len(ids) + 1/2`

Reply:
Not necessarily all graph construction algorithms are working based on a fanout. For example DHT is working based on size of `ids` list and imposing a fanout on that breaks its connectedness. It however _guarantees_ some fanout as the result, inherently. On the other hand, randomized topology construction utilizes a probability `p` to sample the topology. So, I ended up needing a generalized way of representing the interface, and think letting such parameters be set on the constructor of the implementations would be easier to handle and generalize.

Reply:
In any implementation, we will have to do this:
`others := all.Filter(filter.Not(filter.In(shouldHave)))`,
Hence, it will be just simple to make the interface take only one ID list - the one it should choose from. The fanout represent the max value and if the topology cannot keep the return list under that limit then it can error out but not taking a max value means it can potentially return a list of size `n` and we would end up connection to `n` nodes.